src/raster/r.viewshed.exposure/r.viewshed.exposure.py

```diff
@@ -0,0 +1,1409 @@
+                (lreg_shape[0], 1),
+            ),
+            np_dsm + v_elevation,
+        ]
+    )
+
+    # 3. vector VT, adjusted for cell size
+    #    3D array (lreg_shape[0] x lreg_shape[1] x 3)
+    v_vect = np.array(
+        [
+            (v_loc[0] - t_loc[0]) * reg.nsres,
+            (v_loc[1] - t_loc[1]) * reg.ewres,
+            (v_loc[2] - t_loc[2]),
+        ]
+    )
+
+    # 4. projection of vector VT to XZ and YZ plane, adjusted for cell size
+    v_vect_ns = np.array([(v_loc[0] - t_loc[0]) * reg.nsres, v_loc[2] - t_loc[2]])
+
+    v_vect_ew = np.array([(v_loc[1] - t_loc[1]) * reg.ewres, v_loc[2] - t_loc[2]])
+
+    v_vect_ns_unit = v_vect_ns / np.linalg.norm(v_vect_ns, axis=0)
+    v_vect_ew_unit = v_vect_ew / np.linalg.norm(v_vect_ew, axis=0)
+
+    # 5. size of vector VT
+    #    2D array (lreg_shape[0] x lreg_shape[1])
+    v_scal = np.sqrt(v_vect[0] ** 2 + v_vect[1] ** 2 + v_vect[2] ** 2)
+    # replace 0 distance for central pixel by resolution to avoid division by 0
+    # v_scal = np.where(abs(v_scal) < 1e-6, (reg.nsres + reg.ewres) / 2, v_scal)
+    # grass.verbose(v_scal)
+
+    # 6. vector n, its projection to XZ, YZ plane
+    #   1D array [X, Z], [Y, Z]
+    #   already unit vector
+    n_vect_ns_unit = [np.sin(np.radians(t_loc[4])), np.cos(np.radians(t_loc[4]))]
+    n_vect_ew_unit = [np.sin(np.radians(t_loc[3])), np.cos(np.radians(t_loc[3]))]
+
+    # 7. angles beta (ns), theta (ew) (0-90 degrees)
+    #    2D array (lreg_shape[0] x lreg_shape[1])
+    beta = np.arccos(
+        n_vect_ns_unit[0] * v_vect_ns_unit[:][0]
+        + n_vect_ns_unit[1] * v_vect_ns_unit[:][1]
+    )
+    beta = np.where(beta > math.pi / 2, beta - math.pi / 2, beta)
+
+    theta = np.arccos(
+        n_vect_ew_unit[0] * v_vect_ew_unit[:][0]
+        + n_vect_ew_unit[1] * v_vect_ew_unit[:][1]
+    )
+    theta = np.where(theta > math.pi / 2, theta - math.pi / 2, theta)
+
+    # 8. visual magnitude adjusted for distance weight
+    visual_magnitude = (
+        np.cos(beta) * np.cos(theta) * ((reg.nsres * reg.ewres) / (v_scal ** 2))
+    )
+    sys.stderr.flush()
+
+    # 9. Multiply visual magnitude by binary viewshed and weight
+    return visual_magnitude * np_viewshed * t_loc[-1]
+
+
+def sample_raster_with_points(r_map, cat, density, min_d, v_sample, seed):
+    """Random sample exposure source by vector points
+    :param r_map: Raster map to be sampled from
+    :type r_map: string
+    :param cat: Category of raster map to be sampled from
+    :type cat: string
+    :param density: Sampling density
+    :type density: float
+    :param min_d: Minimum distance between sampling points
+    :type min_d: float
+    :param v_sample: Name of output vector map of sampling points
+    :type v_sample: string
+    :param seed: Random seed
+    :param seed: int
+    :return: Name of output vector map of sampling points
+    :rtype: string
+    """
+    # mask categories of raster map to be sampled from
+    grass.run_command(
```

Comment:
I recall we were able to remove the need for using mask in r.sample.category, so I wonder if this would be applicable here as well?

Reply:
Yes, thanks for the hin. This could indeed be solved with a reclass map ...

Reply:
Looking at this again, the differences is the use of `r.random.cells` here, that does not have an input option (in contrast to `r.random` used in `r.sample.category`) So, i think we stick to using the mask for now, unless we rework the sampling more generally (i.e. using only raster and coordinates).

Reply:
Sure, I was not sure how different was this case from r.sample.category.